src/modules/peek/Peek.FilePreviewer/Controls/ShellPreviewHandlerControl.xaml.cs

```diff
@@ -0,0 +1,183 @@
+// Copyright (c) Microsoft Corporation
+// The Microsoft Corporation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Runtime.CompilerServices;
+using Microsoft.UI;
+using Microsoft.UI.Xaml;
+using Microsoft.UI.Xaml.Controls;
+using Peek.FilePreviewer.Previewers;
+using Windows.Win32;
+using Windows.Win32.Foundation;
+using Windows.Win32.UI.Shell;
+using Windows.Win32.UI.WindowsAndMessaging;
+
+namespace Peek.FilePreviewer.Controls
+{
+    public unsafe sealed partial class ShellPreviewHandlerControl : UserControl
+    {
+        // Making this into a DependencyProperty causes a InvalidCastException
+        private IPreviewHandler? _source;
```

Comment:
I am not sure why it would do that, but if the only way to prevent this is to use an observable property then consider adding the attribute [NotifyPropertyChanged] on the ShellPreviewHandlerControl  and directly use the attribute [ObservableProperty] on this _source field. You will be able to use the generated partial method like SourceChanged.